bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusMessageDeserializer.java

```diff
@@ -0,0 +1,110 @@
+
+/**
+ * Class {@link QbusMessageDeserializer} deserializes all json messages from Qbus. Various json
+ * message formats are supported. The format is selected based on the content of the cmd and event json objects.
+ *
+ * @author Koen Schockaert - Initial Contribution
+ *
+ */
+
+@NonNullByDefault
+class QbusMessageDeserializer implements JsonDeserializer<QbusMessageBase> {
+
+    @Override
+    public @Nullable QbusMessageBase deserialize(final JsonElement json, final Type typeOfT,
+            final JsonDeserializationContext context) throws JsonParseException {
+        final JsonObject jsonObject = json.getAsJsonObject();
+
+        try {
+            String cmd = null;
+            String CTD = null;
+
+            if (jsonObject.has("cmd")) {
+                cmd = jsonObject.get("cmd").getAsString();
+            }
+
+            if (jsonObject.has("CTD")) {
+                CTD = jsonObject.get("CTD").getAsString();
+            }
+
+            JsonElement jsonOutputs = null;
+
+            if (jsonObject.has("outputs")) {
+                jsonOutputs = jsonObject.get("outputs");
+            }
+
+            QbusMessageBase message = null;
+
+            if (jsonOutputs != null) {
+                if (jsonOutputs.isJsonObject()) {
+                    message = new QbusMessageMap();
+
+                    Map<String, String> outputs = new HashMap<>();
+                    for (Entry<String, JsonElement> entry : jsonOutputs.getAsJsonObject().entrySet()) {
+                        outputs.put(entry.getKey(), entry.getValue().getAsString());
+                    }
+                    ((QbusMessageMap) message).setOutputs(outputs);
+
+                } else if (jsonOutputs.isJsonArray()) {
+                    JsonArray jsonOutputsArray = jsonOutputs.getAsJsonArray();
+
+                    message = new QbusMessageListMap();
+
+                    List<Map<String, String>> outputsList = new ArrayList<>();
+                    for (int i = 0; i < jsonOutputsArray.size(); i++) {
+                        JsonObject jsonOutputsObject = jsonOutputsArray.get(i).getAsJsonObject();
+
+                        Map<String, String> outputs = new HashMap<>();
+                        for (Entry<String, JsonElement> entry : jsonOutputsObject.entrySet()) {
+                            outputs.put(entry.getKey(), entry.getValue().getAsString());
+                        }
+                        outputsList.add(outputs);
+                    }
+                    ((QbusMessageListMap) message).setOutputs(outputsList);
+                }
+            }
+
+            if (message != null && cmd != null && CTD != null) {
+                message.setCmd(cmd);
+                message.setSn(CTD);
+            } else {
+                throw new JsonParseException("Unexpected Json type");
+            }
+
+            return message;
+
+        } catch (IllegalStateException e) {
+            throw new JsonParseException("Unexpected Json type");
```

Comment:
It's good practice to include the cause in the exception. What throws a `IllegalStateException` anyway?
```suggestion
            throw new JsonParseException("Unexpected Json type", e);
```